### PR TITLE
discv5 integration

### PR DESCRIPTION
**Pull request by wemeetagain**
Draft status until `@chainsafe/discv5` is published
and a libp2p issue is resolved (https://github.com/libp2p/js-libp2p/pull/600)

### CODE REVIEW

File: packages/lodestar/src/network/nodejs/util.ts [hunk 1]
```diff
@@ -31,11 +31,13 @@ export async function loadPeerIdFromJsonFile(path: string): Promise<PeerId> {
...
+  network: Partial<INetworkOptions> = {}
+): Promise<LibP2p> {
   const peerId = await Promise.resolve(peerIdOrPromise);
   const multiaddrs = network.multiaddrs || defaults.multiaddrs;
   const bootnodes = network.bootnodes || defaults.bootnodes;
   const peerInfo = await initializePeerInfo(peerId, multiaddrs);
-  return new NodejsNode({peerInfo, bootnodes: bootnodes});
-}
\ No newline at end of file
+  return new NodejsNode({peerInfo, bootnodes: bootnodes, discv5: network.discv5});
```
Comment (twoeths) [inline packages/lodestar/src/network/nodejs/util.ts:42]: Just a small note: we'll need to put in `ENRForkID` as spec 0.11 PR https://github.com/ethereum/eth2.0-specs/pull/1652/files

Comment (wemeetagain) [reply packages/lodestar/src/network/nodejs/util.ts:42]: :+1: yeah, the enr should be available through `libp2p._discovery.get("discv5").enr`.

I'm thinking of doing this in additional PRs.

File: packages/lodestar/src/network/discovery/discv5.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      for (const enr of enrs) {
+        try {
+          const peerInfo = new PeerInfo(await enr.peerId());
+          const multiaddrTCP = enr.multiaddrTCP;
+          if (multiaddrTCP) {
+            peerInfo.multiaddrs.add(multiaddrTCP);
+          }
+          this.emit("peer", peerInfo);
+        } catch (e) {
+          continue;
```
Comment (mpetrunic) [inline packages/lodestar/src/network/discovery/discv5.ts]: might be useful to log warning with error?

File: packages/lodestar/src/network/discovery/discv5.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    options.bootEnrs.forEach((bootEnr) => this.discv5.addEnr(bootEnr));
+  }
+
+  async start(): Promise<void> {
+    if (this.started) {
+      return;
+    }
+    this.started = true;
+    await this.discv5.start();
+    setTimeout(() => this.findPeers, 1);
```
Comment (mpetrunic) [inline packages/lodestar/src/network/discovery/discv5.ts]: Why wait 1ms before finding peers?

File: packages/lodestar/src/network/discovery/discv5.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    }
+    this.started = false;
+    await this.discv5.stop();
+  }
+
+  async findPeers(): Promise<void> {
+    while (!this.started) {
+      // Search for random nodes
+      // emit discovered on all finds
+      const enrs = await this.discv5.findNode(createNodeId(randombytes(32)));
```
Comment (mpetrunic) [inline packages/lodestar/src/network/discovery/discv5.ts]: Have you considered async iterable here so it yields everytime it find enr? :smile: 
#asynciterableseverywhere

File: packages/lodestar/src/network/discovery/discv5.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+export interface IDiscv5DiscoveryOptions extends IDiscv5DiscoveryInputOptions {
+  peerInfo: PeerInfo;
+}
+
+/**
+ * Discv5Discovery
+ */
+export class Discv5Discovery extends EventEmitter {
```
Comment (mpetrunic) [inline packages/lodestar/src/network/discovery/discv5.ts]: I think we should release this class as libp2p-discv5 or inside @chainsafe/discv5 along with instructions how to add to libp2p stack?

File: packages/lodestar/src/network/discovery/discv5.ts (hunk 4 above)
Comment (wemeetagain) [reply packages/lodestar/src/network/discovery/discv5.ts]: So discv5 has a different event `discovered` that gets emitted for any enrs it finds, but what is returned here is a little different, we're only getting back the closest 16 peers that we were able to connect to as a result of the lookup.

Unfortunately, I don't think we can make it an async iterable.

File: packages/lodestar/src/network/discovery/discv5.ts (hunk 5 above)
Comment (wemeetagain) [reply packages/lodestar/src/network/discovery/discv5.ts]: :+1:

File: packages/lodestar/src/network/discovery/discv5.ts (hunk 3 above)
Comment (wemeetagain) [reply packages/lodestar/src/network/discovery/discv5.ts]: We don't need to, just was using the timeout to run asyncronously. 
Will replace this line with `this.findPeers();`

File: packages/lodestar/test/unit/network/util.ts [hunk 6]
```diff
@@ -9,7 +8,7 @@ export async function createNode(multiaddr: string): Promise<NodejsNode> {
   const peerId = await createPeerId();
   const enr = ENR.createFromPeerId(peerId);
   const randomPort = Math.round(Math.random() * 40000) + 1000;
-  const bindAddr = Multiaddr(`/ip4/127.0.0.1/udp/${randomPort}`);
+  const bindAddr = `/ip4/127.0.0.1/udp/${randomPort}`;
```
Comment (mpetrunic) [inline packages/lodestar/test/unit/network/util.ts:11]: Not sure if works here, but usually if you put port "0" it will take random available port

**Review by mpetrunic — DISMISSED** [review]

**Comment by wemeetagain** [comment]
ready for re-review

**Review by mpetrunic — APPROVED** [review]

**Pull request merged**